pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/RuntimeASTTransformer.groovy

```diff
@@ -50,16 +57,94 @@ import javax.annotation.Nonnull
 
 import static org.codehaus.groovy.ast.tools.GeneralUtils.*
 import static org.jenkinsci.plugins.pipeline.modeldefinition.parser.ASTParserUtils.*
+import static org.objectweb.asm.Opcodes.ACC_FINAL
+import static org.objectweb.asm.Opcodes.ACC_PUBLIC
+import static org.objectweb.asm.Opcodes.ACC_STATIC
 
 /**
  * Transforms a given {@link ModelASTPipelineDef} into the {@link Root} used for actual runtime. Also attaches a
  * {@link ExecutionModelAction} to the run.
  */
 @SuppressFBWarnings(value="SE_NO_SERIALVERSIONID")
 class RuntimeASTTransformer {
-    RuntimeASTTransformer() {
+    ClassNode rootClassNode
+    ModuleNode moduleNode
+    def methodClassNode = [:]
+    def methodCount = [:]
+    def constLookup = [:]
+//    def variableClassNode = null
+//    def variableValues = [:]
+
+
+    RuntimeASTTransformer(ModuleNode moduleNode) {
+        this.moduleNode = moduleNode
+    }
+
+    Expression ctorXFunction(ClassNode type, Expression args) {
+        String typeName = type.nameWithoutPackage;
+
+        if (!methodCount.containsKey(typeName)) {
+            if (!this.moduleNode.classes.find { it.nameWithoutPackage == typeName }) {
+                methodCount[typeName] = 0
+                methodClassNode[typeName] = new ClassNode("DeclarativeRuntime_${typeName}_${Math.abs(type.hashCode())}", ACC_PUBLIC, ClassHelper.make(Object))
+                this.moduleNode.addClass(methodClassNode[typeName])
+            }
+        }
+
+        ClassNode classNode = methodClassNode[typeName]
+
+        String name = "${typeName}_${methodCount[typeName]++}"
+        while (classNode.methods.find {it.name == name }) {
```

Comment:
Took me a bit to figure out what this is doing - if I'm understanding correctly, it's trying various `Foo_1`, `Foo_2`, etc methods to see if it's already added one to settle on a good name? A comment explaining that would be worthwhile.

Reply:
Simplified in latest changes.